flips/20201310-storage-fees.md

```diff
@@ -0,0 +1,316 @@
+// GetStorageCapacity gets storage capacity in bytes on the address.
+GetStorageCapacity(address Address) (value uint64, err error)
+// SetStorageCapacity sets storage capacity to value bytes on the address.
+SetStorageCapacity(address Address, value uint64) err error
+```
+
+In cadence runtime we need to add `storage_capacity` and `storage_used` read-only fields to auth account. **CAUTION** which value does `storage_used` return? `storage_used` before the transaction or at this moment during the transaction.
+
+`SetStorageCapacity(address, value)` Will be exposed to cadence runtime only to the **StorageFeesContract**. TODO: details on how!
+
+### StorageFees Smart Contract
+
+A new `StorageFees` smart contract will be created. It will use the `FlowFees` contract to collect storage fees. It will be used by the `FlowServiceAccount`. The `FlowServiceAccount` Will be used as a wrapper around `StorageFees` to keep all the service functionality in one place.
+
+`StorageFees` smart contract will have the following responsibilities:
+
+- definitions:
+    - `minimumAccountStorage`: defines the minimum amount of `storage_capacity` an account can have
+    - `minimumStorageUnit`: define the minimum unit (chunk) size of storage. Storage can only be bought (refunded) in a multiple of the minimum unit
+    - `flowPerByte`: define the cost of 1 byte of storage in FLOW
+    - `flowPerAccountCreation`: define the cost of purchasing the initial minimum storage
+- functions:
+    - `purchaseMinimumStorageForAddress(payer, address)`:
+        - the payer pays for the minimum storage needed by the address
+        - checks if `storage_capacity` is already present on the address (if so panic)
+        - sets the `storage_capacity` to `minimumAccountStorage` 
+        - using `FlowFees` deducts `flowPerAccountCreation` from the payer
+    - `purchaseStorageForAddress(payer, address, amount)`:
+        - the payer pays for extra storage needed by the address
+        - checks if `storage_capacity` is present on the address (if not so panic)
+        - checks if amount is divisible by `minimumStorageUnit` (if not panic)
+        - increments `storage_capacity` by amount
+        - using `FlowFees` deducts `flowPerByte` x amount from the payer
+    - `refundStorage(account, amount)`:
+        - the account refunds its own purchased storage to its own vault 
+        - checks if `storage_capacity` is present on the address (if not so panic)
+        - checks if amount is divisible by `minimumStorageUnit` (if not panic)
+        - checks if `storage_capacity` - amount >= `minimumAccountStorage` (if not so panic)
+        - decrement `storage_capacity` by amount
+        - using `FlowFees` add `flowPerByte` x amount to the account
+
+### Register migration
+
+First we need to consider that some accounts will be over the 100kB minimum limit. We need to give more storage capacity to these accounts, but the question is how much more.
```

Comment:
I'm actually really comfortable just giving every current account 1F in storage credit, which is 1MB of storage. It's a completely reasonable "expenditure" of Flow tokens, IMO. We should definitely make sure that every account has enough actual FLOW "in escrow" to manage whatever storage limit we give them.

I seem to recall that there might be one or two accounts over that, and we can just given them 5F of storage, or 10F or whatever.

Reply:
Changed the Register Migration chapter to reflect this.